src/backend/distributed/planner/relation_restriction_equivalence.c

```diff
@@ -2051,17 +2051,14 @@ ContextCoversJoinRestriction(JoinRestrictionContext *joinRestrictionContext,
 		 * We check whether the restrictions in joinRestrictionInTest is a subset
 		 * of the restrictions in joinRestrictionInContext in the sense that all the
 		 * restrictions in the latter already exists in the former.
-		 *
-		 * Also, note that list_difference() returns a list that contains all the
-		 * cells in joinRestrictInfoList that are not in inputJoinRestrictInfoList.
-		 * Finally, each element in these lists is a pointer to RestrictInfo
-		 * structure, where equal() function is implemented for the struct.
 		 */
 		List *joinRestrictInfoListInContext =
 			joinRestrictionInContext->joinRestrictInfoList;
 		List *joinRestrictInfoListInTest =
 			joinRestrictionInTest->joinRestrictInfoList;
-		if (LeftListIsSubset(joinRestrictInfoListInContext, joinRestrictInfoListInTest))
+		bool hasOnlyInnerJoin = CurrentPlannerRestrictionContext()->hasOnlyInnerJoin;
```

Comment:
This part seems complicated. Can we split the two checks?
```
		bool joinIsOnTrue = list_length(joinRestrictInfoListInTest) == 0;
		bool hasOnlyInnerJoin = joinRestrictionContext->hasOnlyInnerJoin;
		if (joinIsOnTrue && !hasOnlyInnerJoin)
		{
			/*
			 * If the join doesn't have a restriction (e.g., ON (true)) and
			 * the planner is aware of at least one non-inner JOIN (e.g., outer/semi
			 * joins), we should not eliminiate joinRestrictInfoListInTest. It can
			 * be useful for detecting not supported outer-join checks even if it doesn't
			 * help for colocation checks.
			 */
			continue;
		}
```